### PR TITLE
UpsampleNetwork

### PR DESCRIPTION
**Pull request by jimchen90**
This Upsampling block is part of WaveRNN model. Now the test is to validate the output dimensions of this block. Other tests will be added after other blocks are combined.
Related to #446

Stack:

Add MelResNet Block #705  #751 
**Add Upsampling Block #724**
Add WaveRNN Model #735 
Add example pipeline with WaveRNN #749

### CODE REVIEW

File: torchaudio/models/_wavernn.py [hunk 1]
```diff
@@ -99,9 +99,129 @@ def forward(self, x: Tensor) -> Tensor:
...
+        self.y_scale = y_scale
+
+    def forward(self, x: Tensor) -> Tensor:
+        r"""Pass the input through the _Stretch2d layer.
+
+        Args:
+            x: the input sequence to the _Stretch2d layer (required).
+
+        Shape:
+            - x: :math:`(N, C, S, T)`.
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:142]: nit: remove period at end

File: torchaudio/models/_wavernn.py [hunk 2]
```diff
@@ -99,9 +99,129 @@ def forward(self, x: Tensor) -> Tensor:
...
+            - output: :math:`(N, C, S * y_scale, T * x_scale)`.
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:143]: nit: remove period at end

File: torchaudio/models/_wavernn.py [hunk 3]
```diff
@@ -99,9 +99,129 @@ def forward(self, x: Tensor) -> Tensor:
...
+        where N is the batch size, C is the channel size, S is the number of input sequence,
+        T is the length of input sequence.
+        """
+
+        n, c, s, t = x.size()
+        x = x.unsqueeze(-1).unsqueeze(3)
+        x = x.repeat(1, 1, 1, self.y_scale, 1, self.x_scale)
+        return x.view(n, c, s * self.y_scale, t * self.x_scale)
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:151]: Isn't this redundant? unsqueeze, repeat, view.

File: torchaudio/models/_wavernn.py [hunk 4]
```diff
@@ -99,9 +99,129 @@ def forward(self, x: Tensor) -> Tensor:
...
+            x: the input sequence to the _Stretch2d layer (required).
+
+        Shape:
+            - x: :math:`(N, C, S, T)`.
+            - output: :math:`(N, C, S * y_scale, T * x_scale)`.
+        where N is the batch size, C is the channel size, S is the number of input sequence,
+        T is the length of input sequence.
+        """
+
+        n, c, s, t = x.size()
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:148]: Expecting a four tuple is a little rigid, isn't it?

File: torchaudio/models/_wavernn.py (hunk 3 above)
Comment (vincentqb) [reply torchaudio/models/_wavernn.py:151]: Would something like `return x.repeat(1, 1, self.y_scale, 1).repeat_interleave(1, 1, 1, self.x_scale)` do the same?

File: torchaudio/models/_wavernn.py (hunk 4 above)
Comment (vincentqb) [reply torchaudio/models/_wavernn.py:148]: e.g. some functions support `...` to mean an arbitrary number of dimensions, see functionals in torchaudio.

File: torchaudio/models/_wavernn.py [hunk 5]
```diff
@@ -99,9 +99,129 @@ def forward(self, x: Tensor) -> Tensor:
...
+        self.upsample_layers = nn.Sequential(*up_layers)
+
+    def forward(self, x: Tensor) -> Tensor:
+        r"""Pass the input through the _UpsampleNetwork layer.
+
+        Args:
+            x: the input sequence to the _UpsampleNetwork layer (required).
+
+        Shape:
+            - x: :math:`(N, S, T)`.
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:214]: See notation in [readme](https://github.com/pytorch/audio/blob/master/README.md) for output of spectrogram
> Spectrogram: (channel, time) -> (channel, freq, time)

File: torchaudio/models/_wavernn.py [hunk 6]
```diff
@@ -99,9 +99,129 @@ def forward(self, x: Tensor) -> Tensor:
...
+            - output: :math:`(N, (T - 2 * pad) * Total_Scale, S)`, `(N, (T - 2 * pad) * total_scale, P)`.
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:215]: nit: lower/upper case in `total_scale` and `Total_Scale`. I like the first better :)

File: torchaudio/models/_wavernn.py [hunk 7]
```diff
@@ -99,9 +99,129 @@ def forward(self, x: Tensor) -> Tensor:
...
+        where N is the batch size, S is the number of input sequence, T is the length of input sequence.
+        P is the number of output sequence. Total_Scale is the product of all elements in upsample_scales.
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:217]: nit: `P = output_dims` ? just do that, or specify that `P = output_dims`.

File: torchaudio/models/_wavernn.py [hunk 8]
```diff
@@ -99,9 +99,129 @@ def forward(self, x: Tensor) -> Tensor:
...
+        """
+
+        resnet_output = self.resnet(x).unsqueeze(1)
+        resnet_output = self.resnet_stretch(resnet_output)
+        resnet_output = resnet_output.squeeze(1)
+
+        upsampling_output = self.upsample_layers(x.unsqueeze(1))
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:224]: add a line `x = x.unsqueeze(1)`

File: torchaudio/models/_wavernn.py [hunk 9]
```diff
@@ -103,3 +105,122 @@ def forward(self, x: Tensor) -> Tensor:
...
+
+
+class _UpsampleNetwork(nn.Module):
+    r"""This is an upsample block based on a stack of Conv2d and Strech2d layers.
+    It is a block used in WaveRNN. WaveRNN is based on the paper "Efficient Neural Audio Synthesis".
+    Nal Kalchbrenner, Erich Elsen, Karen Simonyan, Seb Noury, Norman Casagrande, Edward Lockhart,
+    Florian Stimberg, Aaron van den Oord, Sander Dieleman, Koray Kavukcuoglu. arXiv:1802.08435, 2018.
+
+    Args:
+        upsample_scales: the list of upsample scales (required)
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:159]: nit: remove "(required)"

File: torchaudio/models/_wavernn.py (hunk 3 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: You are right. `x.repeat_interleave(y_scale, 2).repeat_interleave(x_scale, 3)` do the same. I tested it in [notebook](https://github.com/jimchen90/TestRep/blob/master/Stretch2d_function_test.ipynb).

File: torchaudio/models/_wavernn.py (hunk 5 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: Variable names have been updated.

File: torchaudio/models/_wavernn.py (hunk 8 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: This line has been added.

File: torchaudio/models/_wavernn.py (hunk 4 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: I have updated it.

File: torchaudio/models/_wavernn.py (hunk 6 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: Yes, `total_scale` looks better. Fixed.

File: torchaudio/models/_wavernn.py (hunk 7 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: This name has been updated. I use `n_output` here to match other places. No single letter is used in docstring now.

File: torchaudio/models/_wavernn.py (hunk 9 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: All "(required)"s in docstring have been removed.

File: torchaudio/models/_wavernn.py (hunk 1 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: Fixed.

Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: Fixed.

File: torchaudio/models/_wavernn.py (hunk 2 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: Fixed.

File: test/test_models.py [hunk 10]
```diff
@@ -1,8 +1,12 @@
...
+
 import torch
-from torchaudio.models import Wav2Letter, _MelResNet
+from torchaudio.models import Wav2Letter, _MelResNet, _UpsampleNetwork
+
+from . import common_utils
 
 
-class TestWav2Letter:
+class TestWav2Letter(common_utils.TorchaudioTestCase):
```
Comment (vincentqb) [inline test/test_models.py:9]: nit: please rebase on master

File: test/test_models.py [hunk 11]
```diff
@@ -31,21 +35,49 @@ def test_mfcc(self):
...
-        pad = 2
+        n_time = 200
+        n_freq = 100
+        n_output = 256
+        n_res_block = 10
+        n_hidden = 128
+        kernel_size = 5
 
-        model = _MelResNet(res_blocks, input_dims, hidden_dims, output_dims, pad)
+        model = _MelResNet(n_res_block, n_freq, n_hidden, n_output, kernel_size)
```
Comment (vincentqb) [inline test/test_models.py:49]: nit: these changes should be in a separate PR

**Comment by vincentqb** [comment]
Is there a doc available? Can you attach the link?

File: torchaudio/models/_wavernn.py [hunk 12]
```diff
@@ -50,56 +47,159 @@ def forward(self, x: Tensor) -> Tensor:
...
+                 n_res_block: int = 10,
+                 n_freq: int = 128,
+                 n_hidden: int = 128,
+                 n_output: int = 128,
+                 kernel_size: int = 5) -> None:
+        super().__init__()
+
+        total_scale = 1
+        for upsample_scale in upsample_scales:
+            total_scale *= upsample_scale
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:168]: Please add an assert or error message checking that `total_scale == hop_lengths`, and document this requirement in docstring.

File: torchaudio/models/_wavernn.py [hunk 13]
```diff
@@ -50,56 +47,159 @@ def forward(self, x: Tensor) -> Tensor:
...
+        r"""
+        Args:
+            x: the input sequence to the _Stretch2d layer
+
+        Shape:
+            - x: :math:`(..., freq, time)`
+            - output: :math:`(..., freq * y_scale, time * x_scale)`
+        """
+
+        return x.repeat_interleave(self.y_scale, 2).repeat_interleave(self.x_scale, 3)
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:131]: This can be made more general by: replacing 2 by -2 and 3 by -1. This way we can keep `...` here.

File: torchaudio/models/_wavernn.py (hunk 12 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py:155]: Because hop_length is a variable only in WaveRNN. I will add an error message in WaveRNN #749 and document this requirement.

File: test/test_models.py (hunk 11 above)
Comment (jimchen90) [reply test/test_models.py:50]: Thanks. I will rebase it on #751 .

File: torchaudio/models/_wavernn.py [hunk 14]
```diff
@@ -85,3 +87,106 @@ def forward(self, specgram: Tensor) -> Tensor:
         """
 
         return self.melresnet_model(specgram)
+
+
+class _Stretch2d(nn.Module):
+    r"""Stretch layer upscales the frequency and time dimensions of a spectrogram.
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:93]: nit: needs rephrasing. Would "Upscales the frequency and time dimensions of a spectrogram." say the same thing?

Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: Updated. I also updated the docstring in upsamplenetwork as:
`r"""Upscale the dimensions of a spectrogram to match waveform.`

File: torchaudio/models/_wavernn.py [hunk 15]
```diff
@@ -85,3 +87,106 @@ def forward(self, specgram: Tensor) -> Tensor:
...
+                             padding=(0, scale),
+                             bias=False)
+            conv.weight.data.fill_(1. / (scale * 2 + 1))
+            up_layers.append(stretch)
+            up_layers.append(conv)
+        self.upsample_layers = nn.Sequential(*up_layers)
+
+    def forward(self, specgram: Tensor) -> Tensor:
+        r"""Pass the input through the _UpsampleNetwork layer.
+        Args:
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:175]: nit: space before

File: torchaudio/models/_wavernn.py [hunk 16]
```diff
@@ -85,3 +87,106 @@ def forward(self, specgram: Tensor) -> Tensor:
...
+
+        resnet_output = self.resnet(specgram).unsqueeze(1)
+        resnet_output = self.resnet_stretch(resnet_output)
+        resnet_output = resnet_output.squeeze(1)
+
+        specgram = specgram.unsqueeze(1)
+        upsampling_output = self.upsample_layers(specgram)
+        upsampling_output = upsampling_output.squeeze(1)[:, :, self.indent:-self.indent]
+
+        return upsampling_output.transpose(1, 2), resnet_output.transpose(1, 2)
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:192]: nit: I'm not sure whether we should change the convention established in other libraries, but I would not transpose here.

File: torchaudio/models/_wavernn.py [hunk 17]
```diff
@@ -85,3 +87,106 @@ def forward(self, specgram: Tensor) -> Tensor:
...
+
+        Return:
+            Tensor shape: (..., n_freq * freq_scale, n_time * time_scale)
+        """
+
+        return specgram.repeat_interleave(self.freq_scale, -2).repeat_interleave(self.time_scale, -1)
+
+
+class _UpsampleNetwork(nn.Module):
+    r"""Upsample block upscales the dimensions of a spectrogram to match waveform.
```
[the remark this answers is not in the document]
Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: In upsampling block, the input is only a spectrogram and the output has same n_time dimension as waveform. After this block, the output will be combined with the waveform as the input of GRU layer.  I agree that the waveform mentioned here will introduce confusion. How about remove `to match waveform`  and like:
`Upscale the dimensions of a spectrogram. `?

File: torchaudio/models/_wavernn.py (hunk 16 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py:192]: How about adding the transpose in two lines before?
```
resnet_output = resnet_output.squeeze(1).transpose()
upsampling_output = upsampling_output.squeeze(1)[:, :, self.indent:-self.indent].transpose()

return upsampling_output, resnet_output
```
Otherwise I has to transpose it in the WaveRNN code because the second dimension has to be same: 
line 288 in #735 as:
```
mels, aux = self.upsample(specgram)
mels = mels.transpose()
aux = aux.transpose()
```

File: torchaudio/models/_wavernn.py [hunk 18]
```diff
@@ -85,3 +87,106 @@ def forward(self, specgram: Tensor) -> Tensor:
...
+                 time_scale: int,
+                 freq_scale: int) -> None:
+        super().__init__()
+
+        self.freq_scale = freq_scale
+        self.time_scale = time_scale
+
+    def forward(self, specgram: Tensor) -> Tensor:
+        r"""Pass the input through the _Stretch2d layer.
+        Args:
```
Comment (vincentqb) [inline torchaudio/models/_wavernn.py:116]: nit: space

File: test/test_models.py [hunk 19]
```diff
@@ -53,3 +53,31 @@ def test_waveform(self):
         out = model(x)
 
         assert out.size() == (n_batch, n_output, n_time - kernel_size + 1)
+
+
+class TestUpsampleNetwork(common_utils.TorchaudioTestCase):
+
+    def test_waveform(self):
+        """Validate the output dimensions of a _UpsampleNetwork block.
```
Comment (vincentqb) [inline test/test_models.py:61]: nit: remove underscore in name here

Comment (vincentqb) [reply test/test_models.py:61]: let's do that all at the end as you suggested offline

File: torchaudio/models/_wavernn.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (vincentqb) [reply torchaudio/models/_wavernn.py]: sounds good

File: torchaudio/models/_wavernn.py (hunk 18 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py:117]: Space is added. Thanks.

File: torchaudio/models/_wavernn.py (hunk 15 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py:177]: Space is added.

File: torchaudio/models/_wavernn.py (hunk 16 above)
Comment (vincentqb) [reply torchaudio/models/_wavernn.py]: As we talked offline, let's move the transpose to WaveRNN instead.

Comment (jimchen90) [reply torchaudio/models/_wavernn.py]: I removed the transpose operation here and it will be added after the [line ](https://github.com/pytorch/audio/pull/735/files#diff-4680098e69e34986c5cca1f1c26b4303R288)of WaveRNN PR

File: torchaudio/models/_wavernn.py (hunk 15 above)
Comment (jimchen90) [reply torchaudio/models/_wavernn.py:177]: Space is added.
